common/views/components/OpeningHours/OpeningHours.js

```diff
@@ -1,41 +1,119 @@
 // @flow
 import {spacing, font} from '../../../utils/classnames';
-import { defaultPlacesOpeningHours as places } from '../../../model/opening-hours';
+import {placesOpeningHours as places} from '../../../model/opening-hours';
+import {Fragment} from 'react';
+import {formatDate} from '../../../utils/format-date';
+import moment from 'moment';
 
 type Props = {|
   id: string,
   extraClasses?: string
-|}
+  |}
+
+function london(d) {
+  // $FlowFixMe
+  return moment.tz(d, 'Europe/London');
+};
+
+// TODO write tests for these functions
+const flattenedExceptionalOpeningDates = places.map((place) => {
+  if (place.openingHours.exceptional) {
+    return place.openingHours.exceptional.map((openingTimes) => {
+      return openingTimes.overrideDate;
+    });
+  }
+})
+  .filter(_ => _)
+  .reduce((prev, curr) => prev && prev.concat(curr));
+
+const uniqueExceptionalOpeningDates = flattenedExceptionalOpeningDates && flattenedExceptionalOpeningDates.sort((a, b) => {
+  if (a && b) {
+    return a.localeCompare(b);
+  } else {
+    return 0;
+  }
+})
+  .filter((item, i, array) => !i || item !== array[i - 1]);
+
+let groupedIndex = 0;
+
+const exceptionalOpeningPeriods = uniqueExceptionalOpeningDates && uniqueExceptionalOpeningDates.reduce((acc, date, i, array) => {
```

Comment:
I wrote an almost functional `groupBy`: https://codepen.io/davidpmccormick/pen/zWwGoV?editors=0010

Mostly just showing off; not suggesting you implement it.

Reply:
Nice 👍 

https://github.com/jashkenas/underscore/blob/1bfc9f1fb811060c745e7025ec55e35f87c1d43b/underscore.js#L431-L447